### PR TITLE
Introduce task actions to support continuous actions and not pure navigation tasks.

### PR DESCRIPTION
**Pull request by mathfac**
## Motivation and Context
Redesign Habitat-API to support continues actions and not pure navigation tasks.
- **New class for Task Actions**:
  - Added registry for actions classes. We use a class not a method for action definition to provide relevant action space that uses Task/Dataset instance's information like number of words for answer vocabulary, resolution of frame etc.
  - Change `step` function interface to support action parameters 
  - Added step function to the task with actions defined as task's methods
  - Provide action space for each type of action
  - Add reset function to Actions 
  - Added teleport action as example of parameterized action.
  - Provide functionality to register several actions of same type using config only
  - Create ActionSpace GYM space to describe action space in elegant way for multiple actions. Propagate action space through env.

- **Moved `Simulator` calls to `EmbodiedTask`**:
  - Move `is_episode_over` functionality to Task class from Simulator, handle capturing finished conditions for different tasks differently
  - Remove simulator `STOP` action
  - Make Task methods reset and step provide observations that include sim observations
  - Add possible actions to config and task
  - Move from SimulatorAction list to defined Tasks in actions, handle non-stop actions
  - Check usage of `*args, **kwargs` for all sensors, action, measures.
  - Use dict as return type for `act` methods.

- **Made `VecEnv` independent from `Env` step signature.**

## How Has This Been Tested
Added additional test for habitat Tasks and fixed old ones.

## Types of changes

- Docs change / refactoring / dependency upgrade
- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have completed my CLA (see **CONTRIBUTING**)
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: habitat/core/env.py [hunk 1]
```diff
@@ -215,12 +215,19 @@ def _update_step_stats(self) -> None:
...
+    ) -> Observations:
         r"""Perform an action in the environment and return observations.
 
         Args:
             action: action (belonging to ``action_space``) to be performed 
-                inside the environment.
+                inside the environment. Action is a name or index of allowed
+                task's action.
+            action_args: action arguments (belonging to action's
+            ``action_space``) to support parametrized and continues actions.
```
Comment (msavva) [inline habitat/core/env.py]: typo: `continues` -> `continuous`

File: habitat/core/registry.py [hunk 2]
```diff
@@ -147,6 +148,20 @@ def register_measure(cls, to_register=None, *, name: Optional[str] = None):
             "measure", to_register, name, assert_type=Measure
         )
 
+    @classmethod
+    def register_task_action(
+        cls, to_register=None, *, name: Optional[str] = None
+    ):
+        r"""Register a measure to registry with key 'name'
```
Comment (msavva) [inline habitat/core/registry.py]: Perhaps a clearer to say `Add a measure in this registry under key 'name'`

File: habitat/core/registry.py [hunk 3]
```diff
@@ -147,6 +148,20 @@ def register_measure(cls, to_register=None, *, name: Optional[str] = None):
...
+
+        Args:
+            name: Key with which the measure will be registered.
+                If None will use the name of the class
```
Comment (msavva) [inline habitat/core/registry.py]: clarify by saying `will use the name of the measure's class`

File: habitat/core/embodied_task.py [hunk 4]
```diff
@@ -131,6 +135,43 @@ def __init__(
         self._sim = sim
         self._dataset = dataset
 
+        self._possible_actions = OrderedDict()
+        for action in config.POSSIBLE_ACTIONS:
+            assert action in registry.mapping["task_action"]
```
Comment (erikwijmans) [inline habitat/core/embodied_task.py]: use `registry. get_task_action` then assert that the result is not `None`

File: habitat/core/embodied_task.py [hunk 5]
```diff
@@ -131,6 +135,43 @@ def __init__(
...
+            self._possible_actions[action] = registry.mapping["task_action"][
+                action
+            ]
```
Comment (erikwijmans) [inline habitat/core/embodied_task.py]: Maybe worth adding something like
```python
for i, v in enumerate(self._possible_actions.value()):
    self._possible_actions[i] = v
```

for cleaner lookup by index later?

File: habitat/core/embodied_task.py [hunk 6]
```diff
@@ -131,6 +135,43 @@ def __init__(
...
+        action_args: Dict[str, Any] = None,
+    ):
+        if action_args is None:
+            action_args = {}
+        if isinstance(action, (int, np.integer)):
+            if action >= len(self._possible_actions):
+                raise ValueError(f"Action index '{action}' is out of range.")
+            action = list(self._possible_actions.keys())[action]
+        assert (
+            action in self._possible_actions.keys()
```
Comment (erikwijmans) [inline habitat/core/embodied_task.py]:
```suggestion
            action in self._possible_actions
```

File: habitat/tasks/nav/nav_task.py [hunk 7]
```diff
@@ -719,6 +720,30 @@ def __init__(
...
+    def look_up(self):
+        return self._sim.step(SimulatorActions.LOOK_UP)
+
+    @registry.register_task_action(name="look_down")
+    def look_down(self):
+        return self._sim.step(SimulatorActions.LOOK_DOWN)
+
+    @registry.register_task_action(name="stop")
+    def stop(self):
+        return self._sim.step(SimulatorActions.STOP)
```
Comment (erikwijmans) [inline habitat/tasks/nav/nav_task.py]: We should probably rename `SimulatorActions` to `Actions` or `BaseActions` as `STOP` isn't a simulator action anymore :)

Comment (mathfac) [reply habitat/tasks/nav/nav_task.py]: @erikwijmans, but `SimulatorActions`is used only as bridge between Simulator Actions and Task Actions. There will be no task action enum anymore. Registry will be used for that. I can remove `STOP` action for HSIM here.

**Comment by mathfac** [comment]
> The `habitat.SimulatorActions` can't be numeric anymore. The number an action is on is entirely dependent on how many actions the task loads.

`habitat.SimulatorActions` is separate notion and is used only in Task action method and will depend on ActionConfiguration activated. My personal point of view that HabitatSimulator actions can be just method on HabitatSim class will be wrapped with correspondent TaskActions.

**Comment by mathfac** [comment]
@msavva and @erikwijmans thank you for the review, addressed the comments. `SimulatorActions` left unchanged for now. Added teleport action as example of parameterized action and added action spec registration for the task action registry.

File: habitat/core/embodied_task.py (hunk 5 above)
Comment (mathfac) [reply habitat/core/embodied_task.py]: @erikwijmans, make sense, but then it tricky to check number of possible actions. Maybe worth to have separate dict with indexes.

File: examples/benchmark.py [hunk 8]
```diff
@@ -16,7 +16,7 @@ def reset(self):
 
     def act(self, observations):
         action = SimulatorActions.MOVE_FORWARD
-        return action
+        return action, None
```
Comment (apsdehal) [inline examples/benchmark.py]: Suggestion: Can we move to dict based input and output system now? Adding more arguments and return items to the signatures is cumbersome and requires doing a lot of changes to make sure there is no regression. Moving to dict based output and input system will remove these concerns. I foresee in future that if we need to add more arguments and return items, we will have to do this again. 

So, I suggest moving to dict based output, input structure. I moved to it in Pythia and life has been a breeze since then.

File: habitat/core/embodied_task.py [hunk 9]
```diff
@@ -131,6 +143,50 @@ def __init__(
...
+        self._possible_actions = OrderedDict()
+        for action in config.POSSIBLE_ACTIONS:
+            action_method = registry.get_task_action(action)
+            self._possible_actions[action] = action_method
+
+    def step(
+        self,
+        action: Union[str, int],
+        episode: Type[Episode],
+        action_args: Dict[str, Any] = None,
```
Comment (apsdehal) [inline habitat/core/embodied_task.py]: Can we add `*args, **kwargs` here and everywhere, to easily add new arguments later in future and easy backward compatibility?

**Comment by mathfac** [comment]
Working on this change I notice that the scope of changes increases as long as other parts of API change. That would be great to break changes in smaller parts, but they are pretty coupled. I updated list of changes required. Please, take a look if it looks reasonable. Most of them are implemented and as soon as tests will be passing I will update the PR.

**Comment by mathfac** [comment]
CI tests are passing, PR is ready for the review @abhiskk, @erikwijmans, @apsdehal.

File: configs/test/habitat_mp3d_eqa_test.yaml [hunk 10]
```diff
@@ -1,5 +1,8 @@
 TASK:
   TYPE: EQA-v0
+  SENSORS: ['QUESTION_SENSOR']
+  POSSIBLE_ACTIONS: ['STOP', 'MOVE_FORWARD', 'TURN_LEFT', 'TURN_RIGHT', 'ANSWER']
```
Comment (erikwijmans) [inline configs/test/habitat_mp3d_eqa_test.yaml]: Shouldn't `ANSWER` and `STOP` be the same action? You only get the try to answer the question once.

File: examples/new_actions.py [hunk 11]
```diff
@@ -126,19 +127,46 @@ def get(self):
...
+        return self._sim.step(habitat.SimulatorActions.STRAFE_RIGHT)
+
+
 def main():
     habitat.SimulatorActions.extend_action_space("STRAFE_LEFT")
     habitat.SimulatorActions.extend_action_space("STRAFE_RIGHT")
 
     config = habitat.get_config(config_paths="configs/tasks/pointnav.yaml")
     config.defrost()
+    # config.TASK.TYPE = "NavTaskWithStrafe-v0"
```
Comment (erikwijmans) [inline examples/new_actions.py]:
```suggestion
```

File: habitat/core/embodied_task.py [hunk 12]
```diff
@@ -140,6 +230,117 @@ def __init__(
...
+                sensor_cfg.TYPE
+            )
+            print("sensor type {}".format(sensor_cfg.TYPE))
+            task_sensors.append(
+                sensor_type(sim=sim, config=sensor_cfg, dataset=dataset)
+            )
+
+        self.sensor_suite = SensorSuite(task_sensors)
+
+        # TODO: move sensors and measures to `_init_entities`
```
Comment (erikwijmans) [inline habitat/core/embodied_task.py]: Does this TODO still need to happen?

File: habitat/core/embodied_task.py [hunk 13]
```diff
@@ -140,6 +230,117 @@ def __init__(
...
+    def step(self, action: Union[int, Dict[str, Any]], episode: Type[Episode]):
+        if "action_args" not in action or action["action_args"] is None:
+            action["action_args"] = {}
+        action_name = action["action"]
+        if isinstance(action_name, (int, np.integer)):
+            if action_name >= len(self.actions):
+                raise ValueError(
+                    f"Action index '{action_name}' is out of range."
+                )
+            action_name = list(self.actions.keys())[action_name]
```
Comment (erikwijmans) [inline habitat/core/embodied_task.py]: Probably doesn't matter, but can we have a `self._action_keys = list(self.actions.keys())` in the `__init__` method and use that here?  Making this on every step seems excessive.

File: habitat/core/embodied_task.py [hunk 14]
```diff
@@ -140,6 +230,117 @@ def __init__(
...
+        self._is_episode_active = self._check_episode_is_active(
+            observations=observations, action=action, episode=episode
+        )
+
+        return observations
+
+    def get_action_name(self, action_index: int):
+        if action_index >= len(self.actions):
+            raise ValueError(f"Action index '{action}' is out of range.")
+        return list(self.actions.keys())[action_index]
```
Comment (erikwijmans) [inline habitat/core/embodied_task.py]: And here `self._action_keys`!

File: habitat/core/embodied_task.py (hunk 13 above)
Comment (erikwijmans) [reply habitat/core/embodied_task.py]: Actually, shouldn't this be `action_name = self.get_action_name(action_name)`

File: habitat/core/env.py [hunk 15]
```diff
@@ -107,7 +107,7 @@ def __init__(
                 **self._task.sensor_suite.observation_spaces.spaces,
             }
         )
-        self.action_space = self._sim.action_space
+        self.action_space = self._task.action_space()
```
Comment (erikwijmans) [inline habitat/core/env.py]: Why the change to a method instead of a property?

File: habitat/datasets/eqa/mp3d_eqa_dataset.py [hunk 16]
```diff
@@ -52,6 +58,19 @@ def __init__(self, config: Config = None) -> None:
         with gzip.open(config.DATA_PATH.format(split=config.SPLIT), "rt") as f:
             self.from_json(f.read())
 
+        questions_vocabulary = {
```
Comment (erikwijmans) [inline habitat/datasets/eqa/mp3d_eqa_dataset.py]: This PR is big enough, but the question stuff here for EQA needs to be fixed.  The questions are a list of tokens, not an individual token themself.

File: habitat/sims/habitat_simulator/habitat_simulator.py [hunk 17]
```diff
@@ -239,23 +234,16 @@ def reset(self):
...
-        self._is_episode_active = True
         return self._sensor_suite.get_observations(sim_obs)
 
     def step(self, action):
-        assert self._is_episode_active, (
-            "episode is not active, environment not RESET or "
-            "STOP action called previously"
-        )
 
         if action == self.index_stop_action:
```
Comment (erikwijmans) [inline habitat/sims/habitat_simulator/habitat_simulator.py]: Do we need references to stop?

File: habitat/tasks/eqa/eqa_task.py [hunk 18]
```diff
@@ -48,77 +45,249 @@ class EQAEpisode(NavigationEpisode):
...
-    def _get_observation(
+    def get_observation(
         self,
         observations: Dict[str, Observations],
         episode: EQAEpisode,
-        **kwargs
+        **kwargs,
     ):
-        return episode.question.question_text
+        return self._dataset.get_questions_vocabulary()[
```
Comment (erikwijmans) [inline habitat/tasks/eqa/eqa_task.py]: Same comment about fixing the question stuff

File: habitat/tasks/nav/nav_task.py [hunk 19]
```diff
@@ -691,35 +720,166 @@ def update_fog_of_war_mask(self, agent_position):
...
+
+    def _get_uuid(self, *args: Any, **kwargs: Any) -> str:
+        return self.name
+
+    def step(self, *args: Any, **kwargs: Any):
+        r"""Update ``_metric``, this method is called from ``Env`` on each
+        ``step``.
+        """
+        self._task.is_stop_called = True
+        return self._sim.get_observations_at()
```
Comment (erikwijmans) [inline habitat/tasks/nav/nav_task.py]: Yep, looks like the stuff about the stop action can be removed from `simulator.step`

File: habitat/tasks/nav/nav_task.py [hunk 20]
```diff
@@ -691,35 +720,166 @@ def update_fog_of_war_mask(self, agent_position):
...
+        rotation: List[float],
+        **kwargs: Any
+    ):
+        r"""Update ``_metric``, this method is called from ``Env`` on each
+        ``step``.
+        """
+
+        if not isinstance(rotation, list):
+            rotation = list(rotation)
+        # if not isinstance(position, list):
```
Comment (erikwijmans) [inline habitat/tasks/nav/nav_task.py]: Delete or uncomment

File: habitat/core/vector_env.py [hunk 21]
```diff
@@ -291,14 +302,15 @@ def step_at(self, index_env: int, action: int):
         self._is_waiting = False
         return results
 
-    def async_step(self, actions: List[int]) -> None:
+    def async_step(self, data: List[Dict[str, Any]]) -> None:
```
Comment (erikwijmans) [inline habitat/core/vector_env.py]: Let's push the backwards compatibility stuff into here and make `async_step` and `step` have the same signature

File: test/test_mp3d_eqa.py [hunk 22]
```diff
@@ -15,9 +15,13 @@
 from habitat.core.embodied_task import Episode
 from habitat.core.logging import logger
 from habitat.datasets import make_dataset
+from habitat.tasks.eqa.eqa_task import AnswerAction
+from habitat.utils.test_utils import sample_non_stop_action
 
 CFG_TEST = "configs/test/habitat_mp3d_eqa_test.yaml"
-CLOSE_STEP_THRESHOLD = 0.028
+CLOSE_STEP_THRESHOLD = 0.03
+# CLOSE_STEP_THRESHOLD = 0.028
```
Comment (erikwijmans) [inline test/test_mp3d_eqa.py]:
```suggestion
```

File: configs/test/habitat_mp3d_eqa_test.yaml (hunk 10 above)
Comment (mathfac) [reply configs/test/habitat_mp3d_eqa_test.yaml]: I was following setup we had for EQA paper, but agree STOP is redundant.

File: habitat/core/embodied_task.py (hunk 12 above)
Comment (mathfac) [reply habitat/core/embodied_task.py]: Already finished in next update.

File: habitat/datasets/eqa/mp3d_eqa_dataset.py (hunk 16 above)
Comment (mathfac) [reply habitat/datasets/eqa/mp3d_eqa_dataset.py]: Yes, EQA is moving to separate PR. It was here to test the change end-to-end.

File: configs/test/habitat_mp3d_eqa_test.yaml (hunk 10 above)
Comment (erikwijmans) [reply configs/test/habitat_mp3d_eqa_test.yaml]: In the setup for the paper, we had STOP == ANSWER.  The agent only every tried to answer the question when stop was called and there were no additional sensor inputs provided after STOP was called (i.e. the last frame wasn't duplicated)

File: habitat/core/embodied_task.py (hunk 12 above)
Comment (erikwijmans) [reply habitat/core/embodied_task.py]: I don't see this?  Or has that update not been pushed yet?

File: habitat/tasks/nav/nav_task.py [hunk 23]
```diff
@@ -455,7 +461,8 @@ def update_metric(self, *args: Any, episode, action, **kwargs: Any):
         )
 
         if (
-            action == self._sim.index_stop_action
+            hasattr(self._task, "is_stop_called")
```
Comment (erikwijmans) [inline habitat/tasks/nav/nav_task.py]: EmbodiedTask should always have an `is_stop_called` field that gets set to `false` on reset and init.  Avoids this clunkiness.

File: habitat/core/embodied_task.py [hunk 24]
```diff
@@ -309,7 +296,7 @@ def step(self, action: Union[int, Dict[str, Any]], episode: Type[Episode]):
                 raise ValueError(
```
Comment (erikwijmans) [inline habitat/core/embodied_task.py]: This check isn't needed, `get_action_name` does it.

File: habitat/tasks/nav/nav_task.py (hunk 23 above)
Comment (mathfac) [reply habitat/tasks/nav/nav_task.py]: @erikwijmans what if your task doesn't have notion of stopping. When success or finished episode is just reach a target to some proximity.

**Comment by mathfac** [comment]
- Removed EQA related changes. 
- Removed `STOP` action for EQA

File: habitat/tasks/nav/nav_task.py (hunk 23 above)
Comment (erikwijmans) [reply habitat/tasks/nav/nav_task.py]: That's fine.  There is nothing that says it ever has to be set true, but the `hasattr(self._task, "is_stop_called")` check before everything is clunky.  One other suggestion:

Add
```
@property
def is_stop_called(self):
    return getattr(self, "_is_stop_called", False)
```

to the EmbodiedTask class.  The actions can then set `self._is_stop_called = True`.

Comment (erikwijmans) [reply habitat/tasks/nav/nav_task.py]: Btw, with the current method, isn't `_task .is_stop_called`, if it exists, always true?  I don't see where it gets set to false or deleted.

File: habitat/tasks/nav/nav_task.py [hunk 25]
```diff
@@ -396,19 +421,27 @@ class SPL(Measure):
     https://arxiv.org/pdf/1807.06757.pdf
     """
 
-    def __init__(self, sim: Simulator, config: Config):
+    def __init__(
+        self,
+        *args: Any,
+        sim: Simulator,
+        task: EmbodiedTask,
```
Comment (erikwijmans) [inline habitat/tasks/nav/nav_task.py]: I think this will end up casing a circular reference i.e. the task holds a reference to the measure and the measure holds a reference to the task, and thus you can't every free the memory used by the task.  (I believe this is also a problem for the actions and the sensors)

There are a couple ways to deal with this:
1. The environment is the owner of actions/sensors/measures -- no circular references here.  Kinda a clunky design tho.
1. Use `weakref.proxy` when giving the measure -- gets rid of the circular here by giving the X a non-owning reference to the task.  This is kinda a weird solution tho.
1. Different API where there is a `task_state` that gets manipulate/populated by actions, then sensors, then measures.  It is passed in as an argument to their step/get/update functions.

The second one feels the most clean.  1 and 2 are both band-aids where the first makes the internals less clean and the second uses a kinda weird python API.

Comment (mathfac) [reply habitat/tasks/nav/nav_task.py]: That's valuable point. I expect generational garbage collector should take care of this as soon as there is no outside reference to the task. I will pass task in `reset` and `step` method as param to avoid this by design.

File: habitat/tasks/nav/nav_task.py (hunk 23 above)
Comment (mathfac) [reply habitat/tasks/nav/nav_task.py]: The Environment class has step count limit after which it will stop the episode. `is_episode_active` is going to be redefined once it will be inheritance.

File: habitat/tasks/nav/nav_task.py (hunk 25 above)
Comment (erikwijmans) [reply habitat/tasks/nav/nav_task.py]: Yeah, looking at this again, this does look like one of cases where the garbage collector can deal with this.
